staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go

```diff
@@ -94,3 +97,18 @@ func AuditEventFrom(ctx context.Context) *audit.Event {
 	ev, _ := ctx.Value(auditKey).(*audit.Event)
 	return ev
 }
+
+func WithAuditAnnotations(parent context.Context, annotations map[string]string) context.Context {
+	if ae := AuditEventFrom(parent); ae != nil {
+		panic("invalid attempt to set audit annotations on context after WithAuditEvent has been called")
+	}
+	return WithValue(parent, auditAnnotationsKey, annotations)
+}
+
+func AuditAnnotationsFrom(ctx context.Context) (map[string]string, bool) {
+	if ae := AuditEventFrom(ctx); ae != nil {
+		panic("invalid attempt to get audit annotations from context after WithAuditEvent has been called")
+	}
+	annotations, ok := ctx.Value(auditAnnotationsKey).(map[string]string)
+	return annotations, ok
```

Comment:
For something like an annotation added by an authenticator, what should be the behavior when the authentication stack is called twice during a TokenReview?

Reply:
oo good point.  They would be ignored in the current implementation, but it would be easy to wire writing of annotations via the actual audit event in the token review REST storage.

However, we would have to prefix (or otherwise disambiguate) the keys to make it clear that they came from the token review storage.

Reply:
Unless there is a way to do this without adding some custom prefix, I would prefer to wait to address any issues with the token review storage in a different PR since the prefix itself will be an API change.

Reply:
The change itself should be straightforward though:

```diff
diff --git pkg/registry/authentication/tokenreview/storage.go pkg/registry/authentication/tokenreview/storage.go
index 368998df258..0b1e55512e3 100644
--- pkg/registry/authentication/tokenreview/storage.go
+++ pkg/registry/authentication/tokenreview/storage.go
@@ -25,6 +25,7 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/audit"
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/registry/rest"
@@ -32,7 +33,10 @@ import (
 	"k8s.io/kubernetes/pkg/apis/authentication"
 )
 
-var badAuthenticatorAuds = apierrors.NewInternalError(errors.New("error validating audiences"))
+var (
+	badAuthenticatorAuds = apierrors.NewInternalError(errors.New("error validating audiences"))
+	auditEventKeyPrefix  = authentication.Resource("tokenreviews").String() + "/rest/"
+)
 
 type REST struct {
 	tokenAuthenticator authenticator.Request
@@ -101,7 +105,7 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 		return nil, badAuthenticatorAuds
 	}
 
-	if resp != nil && resp.User != nil {
+	if ok && err == nil && resp != nil && resp.User != nil {
 		tokenReview.Status.User = authentication.UserInfo{
 			Username: resp.User.GetName(),
 			UID:      resp.User.GetUID(),
@@ -109,9 +113,15 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 			Extra:    map[string]authentication.ExtraValue{},
 		}
 		for k, v := range resp.User.GetExtra() {
-			tokenReview.Status.User.Extra[k] = authentication.ExtraValue(v)
+			tokenReview.Status.User.Extra[k] = v
 		}
 		tokenReview.Status.Audiences = resp.Audiences
+
+		if ae := genericapirequest.AuditEventFrom(ctx); ae != nil {
+			for key, value := range resp.AuditAnnotations {
+				audit.LogAnnotation(ae, auditEventKeyPrefix+key, value)
+			}
+		}
 	}
 
 	return tokenReview, nil
```

Reply:
Sure. Just thinking this through. I'm not sure what we should do here anyway.

Reply:
If something calls the token review API, we can probably assume that it is trying to authenticate some request (or maybe it is doing a `whoami` style call).  Since the token review API lacks context to understand why the API is called, we cannot really generate a "new" audit event about that request flow.  But it does seem reasonable to audit log "I told the caller of this API that this token was for `X` and the authenticator that gave me that information was `Y`."

Reply:
Generally speaking, recording the first write of an annotation key feels least surprising to me.